spec/unit/controllers/v3/service_instance_controller_spec.rb

```diff
@@ -0,0 +1,198 @@
+        expect(response.status).to eq 404
+        expect(response.body).to include('Service instance not found')
+      end
+    end
+
+    context 'when the target space does not exist' do
+      before do
+        req_body[:data] = [{ guid: 'nonexistant-space-guid' }]
+      end
+
+      it 'returns a 422' do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+        expect(response.status).to eq 422
+        expect(response.body).to include('Unable to share to spaces')
+        expect(response.body).to include('nonexistant-space-guid')
+      end
+    end
+
+    context 'when multiple target spaces do not exist' do
+      before do
+        req_body[:data] = [
+          { guid: 'nonexistant-space-guid' },
+          { guid: 'nonexistant-space-guid2' },
+          { guid: target_space.guid }
+        ]
+      end
+
+      it 'does not share to any of the valid spaces and returns a 422' do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+        expect(response.status).to eq 422
+        expect(response.body).to include('Unable to share to spaces')
+        expect(response.body).to include('nonexistant-space-guid')
+        expect(response.body).to include('nonexistant-space-guid2')
+        expect(service_instance.shared_spaces).to_not include(target_space)
+      end
+    end
+
+    context 'when the service instance has already been shared with the specified space' do
+      before do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+      end
+
+      it 'returns a 200 and leaves the existing share intact' do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+
+        expect(response.status).to eq 200
+        expect(service_instance.shared_spaces).to include(target_space)
+      end
+    end
+
+    context 'when the request is malformed' do
+      let(:req_body) {
+        {
+          bork: 'some-name',
+        }
+      }
+
+      it 'returns a 422' do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+        expect(response.status).to eq 422
+      end
+    end
+
+    describe 'permissions by role' do
+      context 'when the user is a space developer in the source space' do
+        before do
+          set_current_user_as_role(role: 'space_developer', org: source_space.organization, space: source_space, user: user)
+        end
+
+        role_to_expected_http_response = {
+          'admin'               => 200,
+          'space_developer'     => 200,
+          'admin_read_only'     => 403,
+          'global_auditor'      => 403,
+          'space_manager'       => 403,
+          'space_auditor'       => 403,
+          'org_manager'         => 403,
+          'org_auditor'         => 422,
+          'org_billing_manager' => 422,
```

Comment:
Why do `org_auditor` and `org_billing_manager` get a 422 when they have a role in the source space but get a 404 in the target space tests below?

Reply:
The rule of thumb we follow is: If they don't have access to a resource in the URL, they get a 404. If they don't have access to a resource in the body, they get a 422.